pkg/api/addons.go

```diff
@@ -432,3 +446,51 @@ func synthesizeAddonsConfig(addons []KubernetesAddon, addon KubernetesAddon, isU
 		addons[i] = assignDefaultAddonVals(addons[i], addon, isUpdate)
 	}
 }
+
+func getKubeDNSAddonContainersSpec(cs *ContainerService, specConfig KubernetesSpecConfig) []KubernetesContainerSpec {
+	k8sComponents := K8sComponentsByVersionMap[cs.Properties.OrchestratorProfile.OrchestratorVersion]
+	spec := []KubernetesContainerSpec{
+		{
+			Name:           "kube-dns-deployment",
+			Image:          specConfig.KubernetesImageBase + k8sComponents["kube-dns"],
+			CPURequests:    "100m",
+			MemoryRequests: "70Mi",
+			MemoryLimits:   "170Mi",
+		},
+		{
+			Name:           "dnsmasq-nanny",
+			Image:          specConfig.KubernetesImageBase + k8sComponents["dnsmasq"],
+			CPURequests:    "150m",
+			MemoryRequests: "20Mi",
+		},
+	}
+
+	if cs.Properties.OrchestratorProfile.NeedsExecHealthz() {
+		execHealthzSpec := KubernetesContainerSpec{
+			Name:           "exechealthz",
+			Image:          specConfig.KubernetesImageBase + k8sComponents["exechealthz"],
+			CPURequests:    "10m",
+			MemoryRequests: "50Mi",
+			MemoryLimits:   "50Mi",
+		}
+		spec = append(spec, execHealthzSpec)
+	} else {
+		sidecarSpec := KubernetesContainerSpec{
+			Name:           "k8s-dns-sidecar",
+			Image:          specConfig.KubernetesImageBase + k8sComponents["k8s-dns-sidecar"],
+			CPURequests:    "150m",
+			MemoryRequests: "20Mi",
+		}
+		spec = append(spec, sidecarSpec)
+	}
+	return spec
+}
+
+func getKubeDNSClusterDomain(k *KubernetesConfig) string {
+	if k.KubeletConfig != nil {
+		if val, ok := k.KubeletConfig["--cluster-domain"]; ok {
+			return val
+		}
+	}
+	return DefaultKubernetesClusterDomain
```

Comment:
I assume the reason we need to do this is because kubelet defaults happen after addon defaults? if so we need to be careful about using this logic since it's a duplicate, if the default logic in kubelet defaults changes we'll need to make sure to update this as well. Could you add a code comment to explain why we need to do this here?

Reply:
The coupling here is between the kubelet runtime `--cluster-domain` configuration, and the `--domain=` value of the `kube-dns-deployment` container runtime in the cluster. The only possible duplication is the "fallback to default" implementation, which would not be necessary if the kubeletConfig "set default if empty" implementation happened prior to this (in that case we'd simply take the value of kubelet's `--cluster-domain` configuration, trusting that it would be valid).

In any event, because we're pulling from a common, known-working default `DefaultKubernetesClusterDomain`, I think this is actually maintenance-safe going forward. The only risk would be carelessly undoing this default value inheritance, say, changing it to a string literal in one place that was different from the const. I think that's unlikely.